### ISSUE TITLE
High Rate of INTERNAL_ERROR from executeTask

### ISSUE THREAD

**Issue by heetchKarakasli** [GoogleCloudPlatform/recaptcha-enterprise-mobile-sdk#28]
**Describe the bug**
I have created ticket for [same issue before](https://github.com/GoogleCloudPlatform/recaptcha-enterprise-mobile-sdk/issues/25). Since it is already closed, I wanted to open new one to share changes after I have upgraded library version from `18.0.0` to `18.0.1`

We have been continue to facing many internal errors for our Android Client when call `executeTask`, about %9 of execution returns with INTERNAL_ERROR as we tracked on **Firebase Crachlytics**.

Before doing that we ignored to users who has potential internet connection issues by checking their authentication results after token exception. End of the 7 days, our numbers are almost fixed to %9, it means %9 of users could not able to get token at least one time a day.

**Integration Method**
It is integrated via Gradle `com.google.android.recaptcha:recaptcha:18.0.1`

This is gradle configuration of our application. 
We are not sure, but this part is not same what documents says, Jetifier is not enabled for our app. 
So it is good to mention about this configuration for possible reasons.

```
android.useAndroidX=true
android.enableJetifier=false
```

**SDK Version (e.g. 18.0.1):** 
18.0.1

**To Reproduce**
Could not detect reproduce steps, we can get same issues time by time for staging or production for any Android Version or and android device brand. There is not any specific relation between android version or brand etc...

**Expected behavior**
We expect to see low ration who could not get token with any reason. We are trying to find reasons of Internal Error.

**Additional context**

**Most frequently**
```
Non-fatal Exception: com.google.android.recaptcha.RecaptchaException: Internal Error
       at com.google.android.recaptcha.internal.zzbf.zzb(com.google.android.recaptcha:recaptcha@@18.0.1:4)
       at com.google.android.recaptcha.internal.zzbf.onReceivedError(com.google.android.recaptcha:recaptcha@@18.0.1:2)
       at android.webkit.WebViewClient.onReceivedError(WebViewClient.java:337)
       at com.google.android.recaptcha.internal.zzbf.onReceivedError(com.google.android.recaptcha:recaptcha@@18.0.1:3)
       at YB0.c(YB0.java:61)
       at Ea.handleMessage(Ea.java:454)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:263)
       at android.app.ActivityThread.main(ActivityThread.java:8292)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:612)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1006)
```

**Second Most Frequently;**
```
Non-fatal Exception: java.lang.Throwable: kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms
       at com.heetch.recaptcha.GoogleRecaptchaProvider.observeToken$lambda$9$lambda$8(GoogleRecaptchaProvider.java:62)
       at com.heetch.recaptcha.GoogleRecaptchaProvider$$InternalSyntheticLambda$5$416df1b00c435c8c730320d6531a4cf05cc44165751598f344685f82a2c342ec$0.onComplete(GoogleRecaptchaProvider.java:30)
       at com.google.android.gms.tasks.zzi.run(zzi.java:1)
       at com.google.android.gms.measurement.internal.zzhh.run$bridge(com.google.android.gms:play-services-measurement-impl@@20.1.0:94)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loopOnce(Looper.java:226)
       at android.os.Looper.loop(Looper.java:313)
       at android.app.ActivityThread.main(ActivityThread.java:8669)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1135)
```

**Third Most Frequently**
```
Non-fatal Exception: com.google.android.recaptcha.RecaptchaException: Internal Error.
       at com.google.android.recaptcha.internal.zzbh.zza(com.google.android.recaptcha:recaptcha@@18.0.1:8)
       at com.google.android.recaptcha.internal.zzbh.zzoid(com.google.android.recaptcha:recaptcha@@18.0.1:5)
       at android.os.MessageQueue.nativePollOnce(MessageQueue.java)
       at android.os.MessageQueue.next(MessageQueue.java:379)
       at android.os.Looper.loop(Looper.java:144)
       at android.os.HandlerThread.run(HandlerThread.java:65)
```

Thanks for your support.

**Comment by mcorner** [comment]
Yes, thanks for the report.  These internal errors are catch-all errors that don't provide a lot of clarity to you, but the overall failure rate does match our internal numbers.  We have another revision coming out (real soon) that will disambiguate the network errors/timouts from internal error, but there are other sources of errors that w are steadily stamping out.

For now, the best strategy is to simply retry, but I would wait for the next version 18.1.0 which fixes a race condition that can be triggered by retrying quickly.

**Comment by MadalinaSpataru** [comment]
Thanks @mcorner  for the quick reply! 2 few questions:
- when will 18.1.0 be available? 
- is it correct to say that in 18.1.0 the failure rate will stay the same, close to 9%? Basically that there is no improvement to the failure rate.

**Comment by mcorner** [comment]
The failure rate is a combination of factors in the SDK and the backend. So you may see some improvement in 18.1.0, but more likely improvements will happen steadily in the backend without notice.

(trying to push 18.1.0 out the door kicking and screaming)

**Comment by heetchKarakasli** [comment]
Thanks for information @mcorner, but I could not resolve the version `18.1.0` when I can resolve `18.0.0` or `18.0.1`. 
`18.1.0` version was announced [here](https://cloud.google.com/recaptcha-enterprise/docs/release-notes) for Android on 1 Feb. Am I missing something?

`Failed to resolve: com.google.android.recaptcha:recaptcha:18.1.0`

**Comment by mcorner** [comment]
yep, the release on android didn't quite make it out the door.  Should be real soon...

**Comment by mcorner** [comment]
Forgot to follow up here, 18.1.1 got released (skipped over 18.1.0).  This is largely the same issue as #21 , but we have already seen a huge improvement in errors from execute on Android.  We are continuing to improve this, but for now we have largely mitigated this.  When encountering errors we do suggest a rety to get over any transient networking issues.
